framework/src/problems/ReferenceResidualProblem.C

```diff
@@ -412,87 +423,61 @@ ReferenceResidualProblem::checkNonlinearConvergence(std::string & msg,
 
     for (unsigned int i = 0; i < _group_soln_var_names.size(); ++i)
     {
-      auto ref_var_name =
-          _reference_vector ? _group_soln_var_names[i] + "_ref" : _group_ref_resid_var_names[i];
-      _console << "   " << std::setw(maxwsv + 2) << std::left << _group_soln_var_names[i] + ":";
+      _console << "   " << std::setw(maxwsv + (_local_norm ? 5 : 2)) << std::left
+               << (_local_norm ? "norm " : "") + _group_soln_var_names[i] + ": ";
 
       if (_group_output_resid[i] == _group_resid[i])
         _console << _group_output_resid[i];
       else
         _console << _group_resid[i] << " (" << _group_output_resid[i] << ')';
-      _console << "  " << std::setw(maxwrv + 2) << ref_var_name + ":" << _group_ref_resid[i]
-               << '\n';
+
+      if (!_local_norm)
+      {
+        const auto ref_var_name =
+            _reference_vector ? _group_soln_var_names[i] + "_ref" : _group_ref_resid_var_names[i];
+        _console << "  " << std::setw(maxwrv + 2) << ref_var_name + ":" << _group_ref_resid[i];
+      }
+      _console << '\n';
     }
 
     _console << std::flush;
   }
+}
 
-  NonlinearSystemBase & system = getNonlinearSystemBase();
-  MooseNonlinearConvergenceReason reason = MooseNonlinearConvergenceReason::ITERATING;
-  std::stringstream oss;
-
-  if (fnorm != fnorm)
-  {
-    oss << "Failed to converge, function norm is NaN\n";
-    reason = MooseNonlinearConvergenceReason::DIVERGED_FNORM_NAN;
-  }
-  else if (it >= _nl_forced_its && fnorm < abstol)
-  {
-    oss << "Converged due to function norm " << fnorm << " < " << abstol << std::endl;
-    reason = MooseNonlinearConvergenceReason::CONVERGED_FNORM_ABS;
-  }
-  else if (nfuncs >= max_funcs)
+bool
+ReferenceResidualProblem::checkRelativeConvergence(const PetscInt it,
+                                                   const Real fnorm,
+                                                   const Real the_residual,
+                                                   const Real rtol,
+                                                   const Real abstol,
+                                                   std::ostringstream & oss)
+{
+  if (checkConvergenceIndividVars(fnorm, abstol, rtol, the_residual))
   {
-    oss << "Exceeded maximum number of function evaluations: " << nfuncs << " > " << max_funcs
-        << std::endl;
-    reason = MooseNonlinearConvergenceReason::DIVERGED_FUNCTION_COUNT;
+    if (_group_resid.size() > 0)
+      oss << "Converged due to function norm " << fnorm << " < relative tolerance (" << rtol
+          << ") or absolute tolerance (" << abstol << ") for all solution variables\n";
+    else
+      oss << "Converged due to function norm " << fnorm << " < relative tolerance (" << rtol
+          << ")\n";
+    return true;
   }
-
-  if (it >= _nl_forced_its && reason == MooseNonlinearConvergenceReason::ITERATING)
```

Comment:
So the failure in the `geomsearch/patch_update_strategy.always-grid-sequencing` test is due to a difference in this line vs FEProblemBase which requires at least one iteration before convergence is checked, i.e.
```
if ((it >= _nl_forced_its) && reason == MooseNonlinearConvergenceReason::ITERATING)
```
I'm not sure why reference residual allowed checking before one iteration, but I will say this has led to a lot of headaches in the past since the initial residual is sometimes below convergence even if the first _forced_ iteration is far away from convergence. I usually rectify that by always forcing at least one residual.

Reply:
Same failure reason is happening for `scalar_transport/test:multiple-species.exo`. I'm inclined to just regold, but would like to have someone else's buyin.

Reply:
I can take a look at the `scalar_transport` failure